external-crates/move/crates/move-compiler/src/shared/mod.rs

```diff
@@ -866,3 +868,124 @@ macro_rules! process_binops {
 }
 
 pub(crate) use process_binops;
+
+//**************************************************************************************************
+// Virtual file system support
+//**************************************************************************************************
+
+/// Same as `IndexedPackagePath` but the path is a virtual file system added/
+/// symbol.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct IndexedVfsPackagePath {
+    pub package: Option<Symbol>,
+    pub path: VfsPath,
+    pub named_address_map: NamedAddressMapIndex,
+}
+
+pub fn vfs_path_from_str(path: String, vfs_path: &VfsPath) -> Result<VfsPath, VfsError> {
+    // we need to canonicalized paths for virtual file systems as some of them (e.g., implementation
+    // of the physical one) cannot handle relative paths
+    fn canonicalize(p: String) -> String {
+        match dunce::canonicalize(&p) {
+            Ok(s) => s.to_string_lossy().to_string(),
+            Err(_) => p,
+        }
+    }
+
+    vfs_path.join(canonicalize(path))
+}
+
+impl IndexedPackagePath {
+    pub fn to_vfs_path(self, vfs_root: &VfsPath) -> Result<IndexedVfsPackagePath, VfsError> {
+        let IndexedPackagePath {
+            package,
+            path,
+            named_address_map,
+        } = self;
+
+        Ok(IndexedVfsPackagePath {
+            package,
+            path: vfs_path_from_str(path.to_string(), vfs_root)?,
+            named_address_map,
+        })
+    }
+}
+
+/// Determine if the virtual path at `vfs_path` exists distinguishing between whether the path did
+/// not exist, or if there were other errors in determining if the path existed.
+pub fn try_exists(vfs_path: &VfsPath) -> VfsResult<bool> {
+    use VfsResult as R;
+    match vfs_path.metadata() {
+        R::Ok(_) => R::Ok(true),
+        R::Err(e) if matches!(e.kind(), &VfsErrorKind::FileNotFound) => R::Ok(false),
+        R::Err(e) => R::Err(e),
+    }
+}
+
+/// - For each directory in `paths`, it will return all files that satisfy the predicate
+/// - Any file explicitly passed in `paths`, it will include that file in the result, regardless
+///   of the file extension
+/// It implements the same functionality as move-command-line-common::files::find_filenames but for
+/// the virtual file system
+pub fn find_filenames<Predicate: FnMut(&VfsPath) -> bool>(
```

Comment:
A shame, but probably necessary. Is there a common trait between VfsPath and Path?

Reply:
I know that APIs look similar but no common trait between the two...

Reply:
Thoughts on a macro?

Reply:
Admittedly I did not write very complicated macros in Rust so perhaps it's easier and looks prettier than in some other languages. Still, I don't think the result would be pretty (or easy to achieve for that matter) Not only there are differences in the API (e.g., how you traverse a directory content), but also some values require conversion on one side but not the other (e.g., via `path_to_string`), some calls are via module syntax one one side and receiver syntax on the other (e.g., `std::fs::metadata(path)` vs. `path.metadata()`), and some calls on one side return `Result` (that needs to be handled) and not on the other (e.g. `is_file`)....

Reply:
Gotcha, if we cannot share the body, could you just move these into external-crates/move/crates/move-command-line-common/src/files.rs ?
That it would be very obvious that if you fix a bug in one, that you need to fix it in the other?